shared/src/main/scala/mlscript/codegen/typescript/TsTypegen.scala

```diff
@@ -286,9 +286,9 @@ final class TsTypegenCodeBuilder {
           val argList = fields
             .map(field => {
               val arg = typegenCtx.termScope.declareRuntimeSymbol("arg")
-              val argType = toTsType(field._2)
+              val argType = toTsType(field._2.out)
```

Comment:
Here and in the Record case, better make sure the field does not have an `in` defined. If it does, and if that's different from the `out`, raise an error. If it's the same, you can make this a mutable field in TS (no `readony` attribute).

Reply:
I've added checks for both record and tuple fields. On making the class field `readonly` I've tried out an example, and it looks like readonly tuples and records still allow changing their fields. So there shouldn't be any case where the `readonly` needs to be removed, because fields are only inside tuples and records.

```ts
class Box {
    readonly x: [number, number] = [1, 2]
    readonly y = { x: 1, y: 2 }
}

let b = new Box()
b.x[0] = 2
// b.y.x = "hello" // can't change
b.y.x = 100
```

Reply:
readonly applies on fields, not their own components. You should be generating `readonly y = { readonly x: 1, readonly y: 2 }` here. There's probably no corresponding syntax for specific tuple components, though.

Reply:
Wow ts is quite granular it does allow a tuple to be defined as readonly. It's granualarity does not extend to each field however.

```ts
let b: readonly [number, number] = [1, 2]

b[1] = 1
```

So the recourse here is to mark the tuple `readonly` if it's all it's fields don't have an `in` value. But a mix of readonly and non-readonly fields will make the whole tuple non-readonly.

Reply:
Sure, though the safer thing to do here would be to _assume_ everything is readonly unless nothing is. This way, TS code shouldn't unexpectedly mutate what MLscript considers immutable... On the other hand, the handling of readonly in TS is positively stupid (you [can](https://www.typescriptlang.org/play?#code/FAYw9gdgzgLgBAJwFxwN6IKYEMAmkA2AnnAB4oQCuAtgEYYJwC+cAvGqSgIxOiSxxQU6MnEq16TVomBQAdCSkAmYEA) upcast a `{ readonly x: number }` into a `{ x: number }` and then mutate it...), s there really aren't any guarantees.

Reply:
Now tuples are readonly unless all their fields are mutable.